### ISSUE TITLE
New-style Function for differentiable and economical backprop

### ISSUE THREAD

**Issue by beam2d** [chainer/chainer#2871]
This is a proposal of introducing a _new style Function_ to Chainer. It replaces the current `Function`. There are two purposes in this proposal.

1. Support **differentiable backprop** (cf. #388).
2. Reduce the computational cost of backprop; when we only need the gradient of x for a function f(x, y), the current API does not allow to skip the gradient computation for y. This proposal also tries to solve it by improving the API design.

I wrote a design document for this proposal here: https://docs.google.com/document/d/1290BiDaep18bjL9ZHs2a6mSLOTcHzbC-OQn1R9oTrHo/edit#. The design is made so that the current implementations of `Function` can basically be used as is. I would like to start implementing changes to the core features and then will work on updating commonly used Functions. After that, I would like to involve other members to update all the Functions.

Any comments/questions/requests are welcome!

**Comment by delta2323** [comment]
Do you have some idea about examples how users write code to compute the second derivative and train NN with them?

**Comment by beam2d** [comment]
For example, the `grad` attribute of a `Variable` becomes a `Variable` (it might be problematic to make this behavior default; in that case, it would be ok to enable this behavior only when an option (or a config) is set). When only gradients w.r.t. few input variables are required, you can write `gxs = grad((y,), xs)`.

**Comment by beam2d** [comment]
I found that it is dangerous to change the type of Variable.grad from ndarray to Variable because it affects existing user code too much. It is possible to add a flag (either global or an option to methods) to switch the type, then the compatibility issue is avoided but it is confusing. It is safer to keep the type of Variable.grad (then Variable.backward() cannot support differentiable backprop) and support differentiable backprop only via the new grad() function.

**Comment by muupan** [comment]
For backward compatibility, how about adding another attribute, say `Variable.grad_as_variable`?

The interface should be designed to make it easier to implement algorithms that require second-order derivatives. For example, MAML (http://arxiv.org/abs/1703.03400) updates the parameter θ as follows:

θ'_i = θ - α ∇_θ L_i(θ)
θ <- θ - β ∇_θ Σ_i L_i(θ'_i)

where L_i is a loss function and α, β are step sizes. That is, MAML tries to minimize the sum of losses L_i, each of which is evaluated after one step of gradient descent on that loss. Do you have any idea which is better to use, `grad()` or `Variable.grad`, to implement this algorithm?

**Comment by beam2d** [comment]
That sounds good. I will consider that design.

**Comment by beam2d** [comment]
Fixed via #2970.
